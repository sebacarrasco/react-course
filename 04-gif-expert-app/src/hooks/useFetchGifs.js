import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
    // Custom hook que se encarga de hacer la request de las imÃ¡genes
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // Importante hacer notar que los efectos no pueden ser async
    useEffect( () => {
        getGifs(category).then(imgs => {
            setState({
                data: imgs,
                loading: false
            });
        });
    }, [category])

    return state;
}