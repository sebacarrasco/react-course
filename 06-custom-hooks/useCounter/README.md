# useCounter Hook

Example of usage with initial value in 5:

```
    const { counter, increment, decrement, reset } = useCounter(5);
```

On the other hand, if no argument is given, the initial value is cero.