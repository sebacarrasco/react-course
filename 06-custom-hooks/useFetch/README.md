# useFetch hook

Example of usage:

```
    const url = 'endpoint/of/an/api.com';
    const { data, loading, error } = useFetch(url);
```