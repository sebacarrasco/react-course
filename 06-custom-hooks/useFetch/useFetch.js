import { useEffect, useState, useRef } from "react";


export const useFetch = ( url ) => {

    // La primera vez que se llama, el componente está montado
    const isMounted = useRef(true);
    
    const [state, setState] = useState(
        {
            data: null,
            loading: true,
            error: null
        }
    );

    useEffect( () => {
        // Cuando se muestre el componente por primera vez no hará nada

        // Cuando se desmonte el componente se cambiará el valor de isMounted
        return () => isMounted.current = false;
    }, []);

    useEffect( () => {


        // Esto es para que aparezca cargando en las siguientes requests
        setState({
            data: null,
            loading: true,
            error: null
        });

        fetch(url)
            .then(response => response.json())
            .then(data => {
                
                // Solo si el componente está montado se actualizará
                // el estado, porque en caso contrario tiraría error.
                if (isMounted.current)
                {
                    setState({
                        loading: false,
                        error: false,
                        data
                    });
                }
            })
            .catch( () => {
                setState({
                    data:null,
                    loading: false,
                    error: "Couldn't load the info"
                })
            });
    }, [url]); // Se ejecutará solo cuando cambia el url

    return state;

};