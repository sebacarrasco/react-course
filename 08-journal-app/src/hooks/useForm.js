import { useState } from "react";

export const useForm = (initialState = {}) => {
    

    const [values, setValues] = useState(initialState);

    const reset = (newFormState = initialState) => {
        setValues(newFormState);
        console.log("hola")
    }

    const handleInputChange = ( { target } ) => {
        setValues({
            ...values,
            [target.name]: target.value
        });
    };

    return [ values, handleInputChange, reset];

};
