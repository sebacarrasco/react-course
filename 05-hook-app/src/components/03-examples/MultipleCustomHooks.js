import React from 'react';
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';
import "../02-useEffect/effects.css";

export const MultipleCustomHooks = () => {

    const { state:counter, increment } = useCounter(1)

    const url = `https://www.breakingbadapi.com/api/quotes/${ counter }`;
    const {data, loading} = useFetch(url);

    // si data es null -> !data es true y !!data es false (author y quote quedan undefined)
    // si data NO es null -> !data es false y !!data es true
    const { author, quote } = !!data && data[0];

    return (
        <div>
            <h3>Breaking Bad quotes</h3> 
            <hr/> 

            {
                loading?
                    (
                        <div className="alert alert-info text-center">
                            Loading...
                        </div>
                    )
                :
                    (
                        <blockquote className="blockquote text-end">
                            <p id="quote-parragraph"> { quote } </p>
                            <footer className="blockquote-footer"> { author } </footer>
                        </blockquote>
                    )
            }

            <button
                className="btn btn-primary"
                onClick={ () => increment() }
            >
                Next quote
            </button>
        </div>
    )
}
