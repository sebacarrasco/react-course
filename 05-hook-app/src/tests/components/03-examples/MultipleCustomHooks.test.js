import React from 'react';
import { shallow } from "enzyme";
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
jest.mock("../../../hooks/useFetch");
jest.mock('../../../hooks/useCounter');

describe('Tests for mulitple custom hooks', () => {

    beforeEach( () => {
        useCounter.mockReturnValue({state: 1, increment: () => {}});
    });
    
    test('should match the snapshot with default values', () => {

        // Es necesario mockear esto porque se usa dentro
        // de <MultipleCustomHooks />
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        
        expect(wrapper).toMatchSnapshot();
    });
    
    test('should show the info returned by the api', () => {
       
        // Vamos a hacer como si ya tuvieramos la informaci√≥n
        useFetch.mockReturnValue({
            data: [{
                author: "Jesse",
                quote: "Mr white"
            }],
            loading: false,
            error: null
        });
        
        const wrapper = shallow(<MultipleCustomHooks />);
        
        // El elemento con clase alerta es el que dice "loading",
        // por lo que no debe aparecer cuando llega la respuesta de la api
        expect(wrapper.find(".alert").exists()).toBe(false);
        expect(wrapper.find("#quote-parragraph").exists()).toBe(true);
        expect(wrapper.find(".blockquote-footer").exists()).toBe(true);

    });
    

});
