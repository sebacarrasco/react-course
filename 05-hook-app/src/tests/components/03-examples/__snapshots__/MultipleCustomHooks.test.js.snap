// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for mulitple custom hooks should match the snapshot with default values 1`] = `
<div>
  <h3>
    Breaking Bad quotes
  </h3>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
